cmd/cmd_env.go

```diff
@@ -0,0 +1,200 @@
+		Short: "Copy files from environment to local directory",
+		Args:  cobra.MinimumNArgs(1),
+		Run:   envCopy,
+	}
+	flags = cmdCopy.PersistentFlags()
+	PersistConfigCommandFlags(flags)
+
+	cmdUninstall := &cobra.Command{
+		Use:   "uninstall",
+		Short: "Uninstall environment",
+		Run:   envUninstall,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "env command [flags]",
+		Short: "Cross-build environment commands",
+		Args:  cobra.OnlyValidArgs,
+		ValidArgs: []string{
+			cmdInstall.Use,
+			cmdBuild.Use,
+			cmdCopy.Use,
+			cmdUninstall.Use,
+		},
+	}
+	cmd.AddCommand(
+		cmdInstall,
+		cmdBuild,
+		cmdCopy,
+		cmdUninstall,
+	)
+	return cmd
+}
+
+// Action for install command.
+func envInstall(cmd *cobra.Command, args []string) {
+	env := loadEnvironment(false)
+	if env.IsInstalled() {
+		exitWithError("Enviroment already installed")
+	}
+	if err := env.Install(); err != nil {
+		exitWithError(fmt.Sprintf("Failed to install environment: %v", err))
+	}
+}
+
+// Action for build command.
+func envBuild(cmd *cobra.Command, args []string) {
+	flags := cmd.Flags()
+	createImage, _ := flags.GetBool("create-image")
+	c := api.NewConfig()
+	configFlags := NewConfigCommandFlags(flags)
+	nightlyFlags := NewNightlyCommandFlags(flags)
+	nanosVersionFlags := NewNanosVersionCommandFlags(flags)
+	buildImageFlags := NewBuildImageCommandFlags(flags)
+	mergeContainer := NewMergeConfigContainer(configFlags, nightlyFlags, nanosVersionFlags, buildImageFlags)
+	err := mergeContainer.Merge(c)
+	if err != nil {
+		exitWithError(err.Error())
+	}
+	env := loadEnvironment(true)
+	if createImage {
+		if c.Program == "" {
+			exitForCmd(cmd, "When creating an image, program must be specified in config file")
+		}
+		if c.RunConfig.Imagename == "" {
+			c.RunConfig.Imagename = c.Program
+		}
+	}
+	tmpRoot := false
+	if c.TargetRoot == "" {
+		if createImage {
+			c.TargetRoot, err = ioutil.TempDir("", "*")
+			if err != nil {
+				exitWithError(err.Error())
+			}
+			tmpRoot = true
+		} else {
+			exitForCmd(cmd, "When not creating an image, target root must be specified "+
+				"(WARNING: any files in the target root folder will be overwritten)")
+		}
+	}
+	if err = env.Boot(); err != nil {
+		log.Errorf("Failed to start environment: %v", err)
+		env = nil
+		goto cleanup
+	}
+	if err = env.RunCommands(args[0]); err != nil {
+		log.Errorf("Failed to build application: %v", err)
+		goto cleanup
+	}
+	if err = env.GetFiles(c, c.TargetRoot); err != nil {
+		log.Errorf("Failed to get application files: %v", err)
+		goto cleanup
+	}
+	if createImage {
+		c.LocalFilesParentDirectory = c.TargetRoot
+		var p api.Provider
+		var ctx *api.Context
+		p, ctx, err = getProviderAndContext(c, "onprem")
+		if err != nil {
+			log.Errorf("Failed to get provider: %v", err)
+			goto cleanup
+		}
+		_, err = p.BuildImage(ctx)
+		if err != nil {
+			log.Errorf("Failed to build image: %v", err)
+		} else {
+			fmt.Printf("On-prem image '%s' created...\n", c.RunConfig.Imagename)
+		}
+	}
+cleanup:
```

Comment:
i'd convert this goto into a defer() that can be put at the top of the function

Reply:
defer doesn't work when os.Exit() is called, so if we want ops to exit with a non-zero status code when an error is encountered (e.g. when the user specifies non-existent files in the config) we can't rely on defer.

Reply:
ah ok, nvr mind then